src/java.base/share/classes/java/lang/StringCoding.java

```diff
@@ -25,15 +26,46 @@
 
 package java.lang;
 
+import jdk.internal.misc.Unsafe;
 import jdk.internal.vm.annotation.IntrinsicCandidate;
 
+import static jdk.internal.misc.Unsafe.ARRAY_BYTE_BASE_OFFSET;
+
 /**
  * Utility class for string encoding and decoding.
  */
 class StringCoding {
+    private static final Unsafe UNSAFE = Unsafe.getUnsafe();
+    private static final long POSITIVE_MASK = 0b1000000_1000000_1000000_1000000_1000000_1000000_1000000_1000000L;
 
     private StringCoding() { }
 
+    public static int countGreaterThanZero(String s) {
+        byte[] value;
+        return countGreaterThanZero(value = s.value(), 0, value.length);
+    }
+
+    /**
+     * Count the number of leading greater than zero bytes in the range.
+     */
+    public static int countGreaterThanZero(byte[] ba, int off, int len) {
+        int limit = off + len;
+        int i = off;
+        for (; i < limit; i += 8) {
+            long v = UNSAFE.getLong(ba, i + ARRAY_BYTE_BASE_OFFSET);
```

Comment:
Since `value` is a `byte[]`, `UNSAFE.getLong` could get bytes outside the array.

Also, that’s not even considering the fact that the address might not even be `long` aligned for `(off % 8) !⁠= 0` or `(off % 8) !⁠= 4`, depending on the array header size (see [JDK‑8139457] and [JDK‑8314882]).

[JDK‑8139457]: https://bugs.openjdk.org/browse/JDK-8139457
[JDK‑8314882]: https://bugs.openjdk.org/browse/JDK-8314882

```suggestion
        for (int end = limit - 7; i < end; i += 8) {
            long v = UNSAFE.getLongUnaligned(ba, i + ARRAY_BYTE_BASE_OFFSET);
```